src/tests/test_method_middleware.py

```diff
@@ -0,0 +1,324 @@
+    class SpecificException(Exception):
+        pass
+
+    def test_method():
+        middleware_requests.append("test_method_called")
+        return True
+
+    class AuthMiddleware():
+        """
+        An example middleware for authentication purposes
+        """
+
+        def user_is_authenticated(self, path):
+            """
+            Simulating an authenticated method to check for
+            no exception cases
+            """
+            return path[-1].name == 'authenticated_method'
+
+        def __call__(self, request, handler):
+            if self.user_is_authenticated(request.path):
+                result = handler(request)
+                return result
+            raise SpecificException
+
+    # Verify custom default middleware
+    class DefaultMiddleware:
+        def __call__(self, request, handler):
+            # We need to get past the DalMethodResolverMiddleware
+            # which sets the `dal_method` attribute to
+            # the method that has to be called
+            request.dal_method = test_method
+            result = handler(request)
+            return result
+
+    dm = DataManager()
+
+    dal = DataAccessLayer(data_manager=dm, middleware=[AuthMiddleware()],
+                          default_middleware=(DefaultMiddleware(),))
+
+    with dm.context():
+        assert dal.fake.authenticated_method()
+        assert middleware_requests[0] == "test_method_called"
+        with pytest.raises(SpecificException):
+            dal.fake.service.method()
+
+
+def test_dal_attribute_access_returns_dal_method_requester():
+    """
+    Verify that accessing an attribute on the DAL that does not
+    exist returns an instance of DeferredDalAttrRequest.
+
+    This is important because this starts building the call chain
+    which is used to resolve methods later.
+    """
+    dm = DataManager()
+    dal = DataAccessLayer(data_manager=dm)
+
+    # Explicitly show that an attribute that is reset returns a
+    # DMR instance
+    dal.real = "real attribute"
+    del dal.real
+    with dm.context():
+        assert isinstance(dal.real, DeferredDalAttrRequest)
+
+    with dm.context():
+        thing = dal.foo.bar.fake
+        thing2 = dal.service.foo.test.example
+        thing3 = dal.method
+        assert isinstance(thing, DeferredDalAttrRequest)
+        assert isinstance(thing2, DeferredDalAttrRequest)
+        assert isinstance(thing3, DeferredDalAttrRequest)
+        thing4 = getattr(dal, 'random')
+        assert isinstance(thing4, DeferredDalAttrRequest)
+
+    # also make sure to verify the error case where a context
+    # has not been started yet.
+    with pytest.raises(AssertionError):
+        dal.foo.bar.fake
+
+
+@pytest.mark.skip(reason="Not implemented yet, but needs to be.")
+def test_dal_getitem_attribute_access():
+    """
+    Note: We need to account for this, but I'm not sure how it will work yet.
+    Consider this test a placeholder for testing that method, which will need
```

Comment:
Will this be done in this PR?

Reply:
No, I don't think so. The intent for this test here was literally just so we didn't forget about implementing and testing it, but I think we should use the existing tests (and fix them up if necessary) instead of adding another test here. I don't know what we could add in terms of value that would be worth doing here, instead of using the other tests.